### PR TITLE
[dist] feat: unified veomni grad norm clipping 

### PR DESCRIPTION
**Pull request by Luosuu**
* Fall back to legacy gradient divide factor API for EP-FSDP modules due to torch 2.7 compatibility issues.
   * In torch 2.8, this legacy API would be redirected to new API: https://github.com/pytorch/pytorch/blob/d38164a545b4a4e4e0cf73ce67173f70574890b6/torch/distributed/fsdp/_fully_shard/_fully_shard.py#L528
   * The original 2.7 API reference: https://github.com/pytorch/pytorch/blob/134179474539648ba7dee1317959529fbd0e7f89/torch/distributed/fsdp/_fully_shard/_fully_shard.py#L460
* Add clip grad norm test in CI
   * The original test for grad norm clipping credit belongs to @ValMystletainn 
   * Added a EP version for GPU.
 * Fix the bug of EP clip grad norm on GPU where we should set gradient divide factor to world size instead of ep_size

### CODE REVIEW

File: veomni/distributed/torch_parallelize.py [hunk 1]
```diff
@@ -347,7 +347,10 @@ def _experts_shard_placement_fn(param):
             # shard expert
             fully_shard(experts_mod, **expert_fsdp_kwargs)
             # average EP grads across EP ranks
-            experts_mod.set_gradient_divide_factor(parallel_state.ep_size)
+            # NOTE: in torch 2.8 and later we should use
+            # experts_mod.set_gradient_divide_factor(parallel_state.ep_size)
+            # but now for torch 2.7 compatability we still use this legacy API
+            experts_mod.set_reduce_scatter_divide_factor(parallel_state.ep_size)
```
Comment (onehaitao) [inline veomni/distributed/torch_parallelize.py]: You can add a version judgment and support different versions simultaneously

File: tests/utils/test_clip_grad_norm.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        basic_modules=[],
+        enable_reentrant=args.train.enable_reentrant,
+        enable_forward_prefetch=args.train.enable_forward_prefetch,
+    )
+
+    for step in range(10):
+        inputs = torch.randn(1, 16).to(device)
+        loss = model(inputs)
+        loss.backward()
+        grad_norm = veomni_clip_grad_norm(model, args.train.max_grad_norm)
```
Comment (piyifan123) [inline tests/utils/test_clip_grad_norm.py]: nit: fixate to a value here.

File: tests/utils/test_clip_grad_norm.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        logger.info_rank0(f"step: {step}, loss: {loss.item()}, grad_norm: {grad_norm}")
```
Comment (piyifan123) [inline tests/utils/test_clip_grad_norm.py]: can we assert grad after/before clipping.

File: veomni/distributed/clip_grad_norm.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+
+def veomni_clip_grad_norm(
+    model, max_norm: float, norm_type: float = 2.0, error_if_nonfinite: bool = False, foreach: bool | None = None
+):
+    parallel_state = get_parallel_state()
+    dp_mode = parallel_state.dp_mode
+    if dp_mode == "fsdp1":
+        grad_norm = fsdp1_clip_grad_norm(model, max_norm, norm_type)
+    elif dp_mode == "fsdp2" and not is_torch_npu_available():
```
Comment (piyifan123) [inline veomni/distributed/clip_grad_norm.py]: nit: add a note to link to the Github issue on why npu need a different path for clipping grad norm in fsdp2.

File: tests/utils/test_clip_grad_norm.py (hunk 3 above)
Comment (Luosuu) [reply tests/utils/test_clip_grad_norm.py]: done

File: tests/utils/test_clip_grad_norm.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        grad_norm_pre_clip = veomni_clip_grad_norm(model, max_grad_norm)
+        # Run the clipper again to measure the norm after the first pass.
+        grad_norm_post_clip = veomni_clip_grad_norm(model, max_grad_norm)
+        logger.info_rank0(
+            f"step: {step}, loss: {loss.item()}, grad_norm_pre_clip: {grad_norm_pre_clip}, "
+            f"grad_norm_post_clip: {grad_norm_post_clip}"
+        )
+        model.zero_grad()
+
+    torch.testing.assert_close(grad_norm_pre_clip, 4.0, atol=1e-6, rtol=1e-6)
```
Comment (piyifan123) [inline tests/utils/test_clip_grad_norm.py]: why not assert every step?

File: tests/utils/test_clip_grad_norm.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+
+logger = helper.create_logger(__name__)
+
+
+@dataclass
+class Argument:
+    train: "TrainingArguments" = field(default_factory=TrainingArguments)
+
+
```
Comment (piyifan123) [inline tests/utils/test_clip_grad_norm.py]: can we merge this file with test_ep_clip_grad_norm as 2 different test cases to reduce redundnacy?

File: tests/utils/test_ep_clip_grad_norm.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        super().__init__()
+        self.bias = torch.nn.Parameter(torch.ones(16), requires_grad=True)
+        self.decoder = ToyMoeDecoderLayer()
+
+    def set_grad(self):
+        for _, p in self.named_parameters(recurse=True):
+            p.grad = torch.ones_like(p)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        loss = (x + self.bias).sum()
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py:53]: does it need to go through decoder layer? if not, can we add comment saying we don't ready need to run forward since we set grad directly.

File: tests/utils/test_ep_clip_grad_norm.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    )
+
+    max_grad_norm = 1.0
+    model.set_grad()
+    grad_norm_pre_clip = veomni_clip_grad_norm(model, max_grad_norm)
+    # Run the clipper again to measure the norm after the first pass.
+    grad_norm_post_clip = veomni_clip_grad_norm(model, max_grad_norm)
+    logger.info_rank0(f"grad_norm_pre_clip: {grad_norm_pre_clip}, grad_norm_post_clip: {grad_norm_post_clip}")
+
+    expected_grad_norm_pre_clip = math.sqrt(16 + 64 * 16 + 64 * 16 * 32)
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py]: nit: add comment to explain this formula for expected grad norm

File: tests/utils/test_ep_clip_grad_norm.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+
+def test_clip_grad_norm_fsdp2_ep8():
+    command = [
+        "torchrun",
+        "--nnodes=1",
+        "--nproc_per_node=4",
+        "--master_port=4321",
+        "tests/utils/test_ep_clip_grad_norm.py",
+        "--train.expert_parallel_size=4",
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py]: ep size does not match test case title. same above.

File: veomni/optim/optimizer.py [hunk 10]
```diff
@@ -400,15 +393,18 @@ def build_ep_fsdp2_optimizer(
...
         non_ep_params: List[torch.nn.Parameter] = []
 
-        for p in model.parameters():
+        for name, p in model.named_parameters():
             if not p.requires_grad:
                 continue
             if DTensor is not None and isinstance(p, DTensor):
                 mesh = getattr(p, "device_mesh", None)
                 names = getattr(mesh, "mesh_dim_names", []) if mesh is not None else []
+                logger.info_rank0(f"param {name} has device_mesh {mesh} with mesh dim names {names}")
```
Comment (piyifan123) [inline veomni/optim/optimizer.py]: remove those

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 11]
```diff
@@ -66,20 +66,30 @@ def ep_fsdp2_clip_grad_norm(
...
+                if q.grad is not None:
+                    q.grad.mul_(scale)
     # Compute and reduce non-EP
     non_ep_total = _fsdp2_reduce_group(
         params=non_ep_params,
         norm_type=norm_type,
         reduce_groups=[("fsdp", fsdp_group)],
     )
-
+    logger.info_rank0(f"non_ep total grad norm: {non_ep_total}")
```
Comment (piyifan123) [inline veomni/distributed/fsdp2/clip_grad_norm.py]: lower the logging levels here.

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 12]
```diff
@@ -66,20 +66,30 @@ def ep_fsdp2_clip_grad_norm(
     non_ep_params: List[torch.nn.Parameter] = [
         p for p in model._ep_param_groups.get("non_ep", []) if p.grad is not None
     ]
-
+    device_type = get_device_type()
+    if device_type == "npu" and ps.ep_enabled and ps.ep_fsdp_size > 1 and ep_params:
```
Comment (piyifan123) [inline veomni/distributed/fsdp2/clip_grad_norm.py]: 1. Add comment to TODO to link to the same Github issue about this so that we can remove it later.
2. Use the standard if npu util.

File: tests/utils/test_ep_clip_grad_norm.py [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+
+class ToyMoeModel(torch.nn.Module):
+    _no_split_modules = ["ToyMoeDecoderLayer"]
+
+    def __init__(self):
+        super().__init__()
+        self.bias = torch.nn.Parameter(torch.ones(16), requires_grad=True)
+        self.decoder = ToyMoeDecoderLayer()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py:52]: nit: add a note that this model would have all graidents as 1.0

File: tests/utils/test_ep_clip_grad_norm.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+        weights_path=None,
+        enable_full_shard=args.train.enable_full_shard,
+        enable_mixed_precision=args.train.enable_mixed_precision,
+        enable_gradient_checkpointing=args.train.enable_gradient_checkpointing,
+        enable_fsdp_offload=args.train.enable_fsdp_offload,
+        basic_modules=[],
+        enable_reentrant=args.train.enable_reentrant,
+        enable_forward_prefetch=args.train.enable_forward_prefetch,
+    )
+    # building a optimizer is necessary to register ep/non-ep params group
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py]: remove this?

File: tests/utils/test_ep_clip_grad_norm.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+
+        logger.info_rank0("manually checking the initial param grads before any clipping")
+        # check them one-by-one
+        for name, param in model.named_parameters():
+            grad = param.grad
+            if grad is None:
+                continue
+            grad_local = grad.to_local() if isinstance(grad, DTensor) else grad
+            logger.info_rank0(f"Before clipping, the local grad for {name}: {grad_local}")
+            expected = 1.0
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py]: add a note to explain why 1.0

File: tests/utils/test_ep_clip_grad_norm.py [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+            torch.testing.assert_close(
+                grad_local,
+                torch.full_like(grad_local, expected),
+                atol=1e-6,
+                rtol=1e-6,
+                msg=f"Gradient mismatch for {name}, which has local shape {grad_local.shape}",
+            )
+        expected_total_grad_norm = math.sqrt(16 + 64 * 16 + 64 * 16 * 32)
+        total_grad_norm_pre_clip = veomni_clip_grad_norm(model, max_grad_norm)
+        # check whether total grad norm meets our expectation
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py:207]: link to TODO here.

File: tests/utils/test_ep_clip_grad_norm.py [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+            logger.info_rank0(f"Before clipping, the local grad for {name}: {grad_local}")
+            expected = 1.0
+            torch.testing.assert_close(
+                grad_local,
+                torch.full_like(grad_local, expected),
+                atol=1e-6,
+                rtol=1e-6,
+                msg=f"Gradient mismatch for {name}, which has local shape {grad_local.shape}",
+            )
+        expected_total_grad_norm = math.sqrt(16 + 64 * 16 + 64 * 16 * 32)
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py]: add a note to explain why this and rs is also averaging so that world size should not affect this..

File: tests/utils/test_ep_clip_grad_norm.py [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+        logger.info_rank0("manually checking the initial param grads before any clipping")
+        # check them one-by-one
+        for name, param in model.named_parameters():
+            grad = param.grad
+            if grad is None:
+                continue
+            grad_local = grad.to_local() if isinstance(grad, DTensor) else grad
+            logger.info_rank0(f"Before clipping, the local grad for {name}: {grad_local}")
+            expected = 1.0
+            torch.testing.assert_close(
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py]: for npu, since the current workaround is not setting the divide factor, why the gradient is also 1.0? we should have #dp of 1.0 grad accumulated, without the divide factor, this should be #dp * 1.0 / (#dp / #ep). i.e., each ep group/rank has #dp/#ep shards.

File: tests/utils/test_ep_clip_grad_norm.py [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+            )
+
+        # go through each param grad one-by-one after clipping to check whether their value meets our expectation
+        clip_coeff = min(max_grad_norm / expected_total_grad_norm, 1.0)
+        for name, param in model.named_parameters():
+            grad = param.grad
+            if grad is None:
+                continue
+            grad_local = grad.to_local() if isinstance(grad, DTensor) else grad
+            logger.info_rank0(f"After clipping, the local grad for {name}: {grad_local}")
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py]: nit: can we have a helper to assert all grad since it's redundant.

File: tests/utils/test_ep_clip_grad_norm.py [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+
+logger = helper.create_logger(__name__)
+
+
+@dataclass
+class Argument:
+    train: "TrainingArguments" = field(default_factory=TrainingArguments)
+
+
+class ToyMoeModel(torch.nn.Module):
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py:35]: nit: add docstring about this model. 

1. all gradients for each fwd/bwd is always 1.0
2. ep does not have a2a, so gradient accumulation over dp is not real.

File: tests/utils/test_ep_clip_grad_norm.py [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+        # * When ep+fsdp2 is enabled, the divide factor for ep params should still be world size (num of data inputs)
+        #   * in implementation, since ep param in VeOmni can only see ep_fsdp dim, so we need to override its divide factor
+        #   * by applying set_gradient_divide_factor(world_size) for EP modules in torch_parallelize
+        # In general, the divide factor for each param should be its num of different input data, which is overall dp size
+
+        # In this test specifically, model forward is unrelated to inputs and the local grad is always 1
+        # * If there is no grad divide factor set, the default grad divide factor is ep_fsdp_size,
+        # * Since we set grad divide factor to world_size (= fsdp_size = ep size * ep_fsdp_size), we expect grad here to be 1/ep_size
+
+        # On NPU, we are missing PreSumMul ReduceOp for set_gradient_divide_factor,
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py:197]: nit: link to the same github issue as a TODO.

File: tests/utils/test_ep_clip_grad_norm.py [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+        # * Now we skip this API so the expected param grad here should still be 1.0 for both ep and non-ep param
+        # * As a result, we need divide the ep_fsdp_size on local grad during clipping to calculate the total norm correctly
+        expected = 1.0
+        ep_expected = 1.0 / ps.ep_size if not IS_NPU_AVAILABLE else 1.0
```
Comment (piyifan123) [inline tests/utils/test_ep_clip_grad_norm.py:201]: explain more on why we need to divide by ep_size: since the model does not have a real ep a2a

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 23]
```diff
@@ -66,20 +66,29 @@ def ep_fsdp2_clip_grad_norm(
     non_ep_params: List[torch.nn.Parameter] = [
         p for p in model._ep_param_groups.get("non_ep", []) if p.grad is not None
     ]
-
+    if IS_NPU_AVAILABLE and ps.ep_enabled and ep_params:
```
Comment (piyifan123) [inline veomni/distributed/fsdp2/clip_grad_norm.py:69]: add TODO to link to github issue.

File: veomni/distributed/parallel_state.py [hunk 24]
```diff
@@ -339,6 +339,18 @@ def ep_enabled(self) -> bool:
...
         return self.ep_fsdp_device_mesh.get_local_rank("ep")
 
+    @property
+    def ep_fsdp_size(self) -> int:
+        assert self.ep_enabled, "ep_fsdp_size is only available when ep is enabled (ep_size > 1)"
+        return self.fsdp_size // self.ep_size
+
+    @property
+    def ep_gradient_divide_factor(self) -> int:
+        assert self.tp_size == 1
```
Comment (piyifan123) [inline veomni/distributed/parallel_state.py:351]: nit: add why tp_size and pp_size should be 1.

File: veomni/distributed/parallel_state.py [hunk 25]
```diff
@@ -339,6 +339,18 @@ def ep_enabled(self) -> bool:
...
+        assert self.pp_size == 1
+        # for ep+fsdp2, the grad divide factor should alwasy be world size (no matter HSDP or not)
```
Comment (piyifan123) [inline veomni/distributed/parallel_state.py]: also add note why sp does not change this.

File: veomni/distributed/torch_parallelize.py [hunk 26]
```diff
@@ -347,7 +347,16 @@ def _experts_shard_placement_fn(param):
             # shard expert
             fully_shard(experts_mod, **expert_fsdp_kwargs)
             # average EP grads across EP ranks
-            experts_mod.set_gradient_divide_factor(parallel_state.ep_size)
+            # NOTE: in torch 2.8 and later we should use
+            # experts_mod.set_gradient_divide_factor(parallel_state.ep_size)
+            # but now for torch 2.7 compatability we still use this legacy API
+            # Note that NPU does not support PreSumMul so we skip this call
+            # see https://github.com/ByteDance-Seed/VeOmni/issues/241
```
Comment (piyifan123) [inline veomni/distributed/torch_parallelize.py]: nit: TODO(github issue link): due to npu ..., delete this after npu support divide factor...

**Review by piyifan123 — APPROVED** [review]
